File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/CucumberExpressionParser.java
package io.cucumber.cucumberexpressions;

import io.cucumber.cucumberexpressions.Ast.Node;
import io.cucumber.cucumberexpressions.Ast.Token;
import io.cucumber.cucumberexpressions.Ast.Token.Type;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static io.cucumber.cucumberexpressions.Ast.Node.Type.ALTERNATION_NODE;
import static io.cucumber.cucumberexpressions.Ast.Node.Type.ALTERNATIVE_NODE;
import static io.cucumber.cucumberexpressions.Ast.Node.Type.EXPRESSION_NODE;
import static io.cucumber.cucumberexpressions.Ast.Node.Type.OPTIONAL_NODE;
import static io.cucumber.cucumberexpressions.Ast.Node.Type.PARAMETER_NODE;
import static io.cucumber.cucumberexpressions.Ast.Node.Type.TEXT_NODE;
import static io.cucumber.cucumberexpressions.Ast.Token.Type.ALTERNATION;
import static io.cucumber.cucumberexpressions.Ast.Token.Type.BEGIN_OPTIONAL;
import static io.cucumber.cucumberexpressions.Ast.Token.Type.BEGIN_PARAMETER;
import static io.cucumber.cucumberexpressions.Ast.Token.Type.END_OF_LINE;
import static io.cucumber.cucumberexpressions.Ast.Token.Type.END_OPTIONAL;
import static io.cucumber.cucumberexpressions.Ast.Token.Type.END_PARAMETER;
import static io.cucumber.cucumberexpressions.Ast.Token.Type.START_OF_LINE;
import static io.cucumber.cucumberexpressions.Ast.Token.Type.WHITE_SPACE;
import static io.cucumber.cucumberexpressions.CucumberExpressionException.createAlternationNotAllowedInOptional;
import static io.cucumber.cucumberexpressions.CucumberExpressionException.createInvalidParameterTypeName;
import static io.cucumber.cucumberexpressions.CucumberExpressionException.createMissingEndToken;
import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;

final class CucumberExpressionParser {

    /*
     * text := whitespace | ')' | '}' | .
     */
    private static final Parser textParser = (expression, tokens, current) -> {
        Token token = tokens.get(current);
        switch (token.type) {
            case WHITE_SPACE:
            case TEXT:
            case END_PARAMETER:
            case END_OPTIONAL:
                return new Result(1, new Node(TEXT_NODE, token.start(), token.end(), token.text));
            case ALTERNATION:
                throw createAlternationNotAllowedInOptional(expression, token);
            case BEGIN_PARAMETER:
            case START_OF_LINE:
            case END_OF_LINE:
            case BEGIN_OPTIONAL:
            default:
                // If configured correctly this will never happen
                return new Result(0);
        }
    };

    /*
     * name := whitespace | .
     */
    private static final Parser nameParser = (expression, tokens, current) -> {
        Token token = tokens.get(current);
        switch (token.type) {
            case WHITE_SPACE:
            case TEXT:
                return new Result(1, new Node(TEXT_NODE, token.start(), token.end(), token.text));
            case BEGIN_OPTIONAL:
            case END_OPTIONAL:
            case BEGIN_PARAMETER:
            case END_PARAMETER:
            case ALTERNATION:
                throw createInvalidParameterTypeName(token, expression);
            case START_OF_LINE:
            case END_OF_LINE:
            default:
                // If configured correctly this will never happen
                return new Result(0);
        }
    };

    /*
     * parameter := '{' + name* + '}'
     */
    private static final Parser parameterParser = parseBetween(
            PARAMETER_NODE,
            BEGIN_PARAMETER,
            END_PARAMETER,
            singletonList(nameParser)
    );

    /*
     * optional := '(' + option* + ')'
     * option := optional | parameter | text
     */
    private static final Parser optionalParser;
    static {
        List<Parser> parsers = new ArrayList<>();
        optionalParser = parseBetween(
                OPTIONAL_NODE,
                BEGIN_OPTIONAL,
                END_OPTIONAL,
                parsers
        );
        parsers.addAll(asList(optionalParser, parameterParser, textParser));
    }

    /*
     * alternation := alternative* + ( '/' + alternative* )+
     */
    private static final Parser alternativeSeparator = (expression, tokens, current) -> {
        if (!lookingAt(tokens, current, ALTERNATION)) {
            return new Result(0);
        }
        Token token = tokens.get(current);
        return new Result(1, new Node(ALTERNATIVE_NODE, token.start(), token.end(), token.text));
    };

    private static final List<Parser> alternativeParsers = asList(
            alternativeSeparator,
            optionalParser,
            parameterParser,
            textParser
    );

    /*
     * alternation := (?<=left-boundary) + alternative* + ( '/' + alternative* )+ + (?=right-boundary)
     * left-boundary := whitespace | } | ^
     * right-boundary := whitespace | { | $
     * alternative: = optional | parameter | text
     */
    private static final Parser alternationParser = (expression, tokens, current) -> {
        int previous = current - 1;
        if (!lookingAtAny(tokens, previous, START_OF_LINE, WHITE_SPACE, END_PARAMETER)) {
            return new Result(0);
        }

        Result result = parseTokensUntil(expression, alternativeParsers, tokens, current, WHITE_SPACE, END_OF_LINE, BEGIN_PARAMETER);
        int subCurrent = current + result.consumed;
        if (result.ast.stream().noneMatch(astNode -> astNode.type() == ALTERNATIVE_NODE)) {
            return new Result(0);
        }

        int start = tokens.get(current).start();
        int end = tokens.get(subCurrent).start();
        // Does not consume right hand boundary token
        return new Result(result.consumed,
                new Node(ALTERNATION_NODE, start, end, splitAlternatives(start, end, result.ast)));
    };

    /*
     * cucumber-expression :=  ( alternation | optional | parameter | text )*
     */
    private static final Parser cucumberExpressionParser = parseBetween(
            EXPRESSION_NODE,
            START_OF_LINE,
            END_OF_LINE,
            asList(
                    alternationParser,
                    optionalParser,
                    parameterParser,
                    textParser
            )
    );

    Node parse(String expression) {
        CucumberExpressionTokenizer tokenizer = new CucumberExpressionTokenizer();
        List<Token> tokens = tokenizer.tokenize(expression);
        Result result = cucumberExpressionParser.parse(expression, tokens, 0);
        return result.ast.get(0);
    }

    private interface Parser {
        Result parse(String expression, List<Token> tokens, int current);

    }

    private static final class Result {
        final int consumed;
        final List<Node> ast;

        private Result(int consumed, Node... ast) {
            this(consumed, Arrays.asList(ast));
        }

        private Result(int consumed, List<Node> ast) {
            this.consumed = consumed;
            this.ast = ast;
        }

    }

    private static Parser parseBetween(
            Node.Type type,
            Type beginToken,
            Type endToken,
            List<Parser> parsers) {
        return (expression, tokens, current) -> {
            if (!lookingAt(tokens, current, beginToken)) {
                return new Result(0);
            }
            int subCurrent = current + 1;
            Result result = parseTokensUntil(expression, parsers, tokens, subCurrent, endToken, END_OF_LINE);
            subCurrent += result.consumed;

            // endToken not found
            if (!lookingAt(tokens, subCurrent, endToken)) {
                throw createMissingEndToken(expression, beginToken, endToken, tokens.get(current));
            }
            // consumes endToken
            int start = tokens.get(current).start();
            int end = tokens.get(subCurrent).end();
            return new Result(subCurrent + 1 - current, new Node(type, start, end, result.ast));
        };
    }

    private static Result parseTokensUntil(
            String expression,
            List<Parser> parsers,
            List<Token> tokens,
            int startAt,
            Type... endTokens) {
        int current = startAt;
        int size = tokens.size();
        List<Node> ast = new ArrayList<>();
        while (current < size) {
            if (lookingAtAny(tokens, current, endTokens)) {
                break;
            }

            Result result = parseToken(expression, parsers, tokens, current);
            if (result.consumed == 0) {
                // If configured correctly this will never happen
                // Keep to avoid infinite loops
                throw new IllegalStateException("No eligible parsers for " + tokens);
            }
            current += result.consumed;
            ast.addAll(result.ast);
        }
        return new Result(current - startAt, ast);
    }

    private static Result parseToken(String expression, List<Parser> parsers,
            List<Token> tokens,
            int startAt) {
        for (Parser parser : parsers) {
            Result result = parser.parse(expression, tokens, startAt);
            if (result.consumed != 0) {
                return result;
            }
        }
        // If configured correctly this will never happen
        throw new IllegalStateException("No eligible parsers for " + tokens);
    }

    private static boolean lookingAtAny(List<Token> tokens, int at, Type... tokenTypes) {
        for (Type tokeType : tokenTypes) {
            if (lookingAt(tokens, at, tokeType)) {
                return true;
            }
        }
        return false;
    }

    private static boolean lookingAt(List<Token> tokens, int at, Type token) {
        if (at < 0) {
            // If configured correctly this will never happen
            // Keep for completeness
            return token == START_OF_LINE;
        }
        if (at >= tokens.size()) {
            return token == END_OF_LINE;
        }
        return tokens.get(at).type == token;
    }

    private static List<Node> splitAlternatives(int start, int end, List<Node> alternation) {
        List<Node> separators = new ArrayList<>();
        List<List<Node>> alternatives = new ArrayList<>();
        List<Node> alternative = new ArrayList<>();
        for (Node n : alternation) {
            if (ALTERNATIVE_NODE.equals(n.type())) {
                separators.add(n);
                alternatives.add(alternative);
                alternative = new ArrayList<>();
            } else {
                alternative.add(n);
            }
        }
        alternatives.add(alternative);

        return createAlternativeNodes(start, end, separators, alternatives);
    }

    private static List<Node> createAlternativeNodes(int start, int end, List<Node> separators, List<List<Node>> alternatives) {
        List<Node> nodes = new ArrayList<>();
        for (int i = 0; i < alternatives.size(); i++) {
            List<Node> n = alternatives.get(i);
            if (i == 0) {
                Node rightSeparator = separators.get(i);
                nodes.add(new Node(ALTERNATIVE_NODE, start, rightSeparator.start(), n));
            } else if (i == alternatives.size() - 1) {
                Node leftSeparator = separators.get(i - 1);
                nodes.add(new Node(ALTERNATIVE_NODE, leftSeparator.end(), end, n));
            } else {
                Node leftSeparator = separators.get(i - 1);
                Node rightSeparator = separators.get(i);
                nodes.add(new Node(ALTERNATIVE_NODE, leftSeparator.end(), rightSeparator.start(), n));
            }
        }
        return nodes;
    }

}

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/cucumber_expression_parser.rb
require 'cucumber/cucumber_expressions/ast'
require 'cucumber/cucumber_expressions/errors'
require 'cucumber/cucumber_expressions/cucumber_expression_tokenizer'


module Cucumber
  module CucumberExpressions
    class CucumberExpressionParser
      def parse(expression)
        # text := token
        parse_text = lambda do |_, tokens, current|
          token = tokens[current]
          return 1, [Node.new(NodeType::TEXT, nil, token.text, token.start, token.end)]
        end

        # parameter := '{' + text* + '}'
        parse_parameter = parse_between(
            NodeType::PARAMETER,
            TokenType::BEGIN_PARAMETER,
            TokenType::END_PARAMETER,
            [parse_text]
        )

        # optional := '(' + option* + ')'
        # option := parameter | text
        parse_optional = parse_between(
            NodeType::OPTIONAL,
            TokenType::BEGIN_OPTIONAL,
            TokenType::END_OPTIONAL,
            [parse_parameter, parse_text]
        )

        # alternation := alternative* + ( '/' + alternative* )+
        parse_alternative_separator = lambda do |_, tokens, current|
          unless looking_at(tokens, current, TokenType::ALTERNATION)
            return 0, nil
          end
          token = tokens[current]
          return 1, [Node.new(NodeType::ALTERNATIVE, nil, token.text, token.start, token.end)]
        end

        alternative_parsers = [
            parse_alternative_separator,
            parse_optional,
            parse_parameter,
            parse_text,
        ]

        # alternation := (?<=left-boundary) + alternative* + ( '/' + alternative* )+ + (?=right-boundary)
        # left-boundary := whitespace | } | ^
        # right-boundary := whitespace | { | $
        # alternative: = optional | parameter | text
        parse_alternation = lambda do |expression, tokens, current|
          previous = current - 1
          unless looking_at_any(tokens, previous, [TokenType::START_OF_LINE, TokenType::WHITE_SPACE, TokenType::END_PARAMETER])
            return 0, nil
          end

          consumed, ast = parse_tokens_until(expression, alternative_parsers, tokens, current, [TokenType::WHITE_SPACE, TokenType::END_OF_LINE, TokenType::BEGIN_PARAMETER])
          sub_current = current + consumed
          unless ast.map { |astNode| astNode.type }.include? NodeType::ALTERNATIVE
            return 0, nil
          end

          start = tokens[current].start
          _end = tokens[sub_current].start
          # Does not consume right hand boundary token
          return consumed, [Node.new(NodeType::ALTERNATION, split_alternatives(start, _end, ast), nil, start, _end)]
        end

        #
        # cucumber-expression :=  ( alternation | optional | parameter | text )*
        #
        parse_cucumber_expression = parse_between(
            NodeType::EXPRESSION,
            TokenType::START_OF_LINE,
            TokenType::END_OF_LINE,
            [parse_alternation, parse_optional, parse_parameter, parse_text]
        )

        tokenizer = CucumberExpressionTokenizer.new
        tokens = tokenizer.tokenize(expression)
        _, ast = parse_cucumber_expression.call(expression, tokens, 0)
        ast[0]
      end

      private

      def parse_between(type, begin_token, end_token, parsers)
        lambda do |expression, tokens, current|
          unless looking_at(tokens, current, begin_token)
            return 0, nil
          end
          sub_current = current + 1
          consumed, ast = parse_tokens_until(expression, parsers, tokens, sub_current, [end_token])
          sub_current += consumed

          # endToken not found
          unless looking_at(tokens, sub_current, end_token)
            raise MissingEndToken.new(expression, begin_token, end_token, tokens[current])
          end
          # consumes endToken
          start = tokens[current].start
          _end = tokens[sub_current].end
          consumed = sub_current + 1 - current
          ast = [Node.new(type, ast, nil, start, _end)]
          return consumed, ast
        end
      end

      def parse_token(expression, parsers, tokens, start_at)
        for parser in parsers do
          consumed, ast = parser.call(expression, tokens, start_at)
          unless consumed == 0
            return consumed, ast
          end
        end
        # If configured correctly this will never happen
        raise 'No eligible parsers for ' + tokens
      end

      def parse_tokens_until(expression, parsers, tokens, start_at, end_tokens)
        current = start_at
        size = tokens.length
        ast = []
        while current < size do
          if looking_at_any(tokens, current, end_tokens)
            break
          end
          consumed, sub_ast = parse_token(expression, parsers, tokens, current)
          if consumed == 0
            # If configured correctly this will never happen
            # Keep to avoid infinite loops
            raise 'No eligible parsers for ' + tokens
          end
          current += consumed
          ast += sub_ast
        end
        [current - start_at, ast]
      end

      def looking_at_any(tokens, at, token_types)
        for token_type in token_types
          if looking_at(tokens, at, token_type)
            return true
          end
        end
        false
      end

      def looking_at(tokens, at, token)
        if at < 0
          # If configured correctly this will never happen
          # Keep for completeness
          return token == TokenType::START_OF_LINE
        end
        if at >= tokens.length
          return token == TokenType::END_OF_LINE
        end
        tokens[at].type == token
      end

      def split_alternatives(start, _end, alternation)
        separators = []
        alternatives = []
        alternative = []
        alternation.each { |n|
          if NodeType::ALTERNATIVE == n.type
            separators.push(n)
            alternatives.push(alternative)
            alternative = []
          else
            alternative.push(n)
          end
        }
        alternatives.push(alternative)
        create_alternative_nodes(start, _end, separators, alternatives)
      end

      def create_alternative_nodes(start, _end, separators, alternatives)
        nodes = []
        alternatives.each_with_index do |n, i|
          if i == 0
            right_separator = separators[i]
            nodes.push(Node.new(NodeType::ALTERNATIVE, n, nil, start, right_separator.start))
          elsif i == alternatives.length - 1
            left_separator = separators[i - 1]
            nodes.push(Node.new(NodeType::ALTERNATIVE, n, nil, left_separator.end, _end))
          else
            left_separator = separators[i - 1]
            right_separator = separators[i]
            nodes.push(Node.new(NodeType::ALTERNATIVE, n, nil, left_separator.end, right_separator.start))
          end
        end
        nodes
      end
    end
  end
end
